### ISSUE TITLE
Need to support transformer decode into specific candidates

### ISSUE THREAD

**Issue by cmxcn** [tensorflow/models#9028]
# Prerequisites

Please answer the following question for yourself before submitting an issue.

- [N] I checked to make sure that this feature has not been requested already.

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/tree/master/official/nlp/transformer

## 2. Describe the feature you request

In many cases, we need to require the generated sentence of transformer in a specific candidate set. Basically this can be achieved by adding a prefix tree like TRIE and check the result in every decoding step. However I found it difficult to incorporate such function into TF based beam search. Any thoughts about this? Thanks in advance.

## 3. Additional context

Add any other context about the feature request here.

**Comment by saberkun** [comment]
@cmxcn This looks a hard question to me. I am helping consulting TF team internally.
The beam search ops runs in-graph and on-device.
If you are running the beam search on CPU, you probably can add customized ops for the trie.
If you are running it on GPU/TPU, we don't have much idea now.
I know some work breaks the TF logits computation and the beam search loop in C++ server.

**Comment by cmxcn** [comment]
Thanks for the quick response. 
The beam search in my case is running on CPU. 
Looking forward to your info from the TF team.

**Comment by cmxcn** [comment]
Hi, Hongkun, any progress about this?

**Comment by saberkun** [comment]
Sorry, I do not get an easy solution at this moment to make the trie in-graph. I think you will need a customized op holding the resource of trie like the lookup table. 
We do not have cycle to implement this at this moment.
Did you give a try to use a hashtable instead of trie?

Alternatively, you can call session run for each step inside C++ and handle the candidate selection directly inside C++. After selecting the candidates inside beam, you can send the inputs back to the next-step session together with cache tensors, etc.

**Comment by cmxcn** [comment]
Thanks for the response. I'll try the second one. 👍 Some performance degrading with a little dev efforts.
